### PR TITLE
Make PLY parser more tolerant towards files not adhering to sta…

### LINKED ISSUE #3487: PLYReader fails to read without extra line at the end nor comment
## Your Environment
* Operating System and version: Ubuntu 18.04
* Compiler: GCC 7 
* PCL Version: Latest master

## Context
PLYReader fails to read ply file in a few simple scenarios.

## Code to Reproduce
```
int main(){
  pcl::PointCloud<pcl::PointXYZRGB> cld;

  pcl::PLYReader Reader;
  Reader.read("test1.ply", cld); // success
  // Reader.read("test2.ply", cld); // fail
  // Reader.read("test3.ply", cld); // fail
  // Reader.read("test4.ply", cld); // fail

  return 0;
}
```

test1.ply is the only scenario that it will read successfully
```
ply
format ascii 1.0
comment PCL generated
element vertex 3
property float x
property float y
property float z
property uchar red
property uchar green
property uchar blue
element camera 1
property float view_px
property float view_py
property float view_pz
property float x_axisx
property float x_axisy
property float x_axisz
property float y_axisx
property float y_axisy
property float y_axisz
property float z_axisx
property float z_axisy
property float z_axisz
property float focal
property float scalex
property float scaley
property float centerx
property float centery
property int viewportx
property int viewporty
property float k1
property float k2
end_header
127.97639 -267.0896 -900.87177 66 107 172
128.3494 -267.11069 -900.89008 65 103 172
128.06009 -267.20074 -901.63428 68 120 175
0 0 0 1 0 0 0 1 0 0 0 1 0 0 0 0 0 3 1 0 0

```

If we simply delete the last empty line from test1.ply to become test2.ply it fails
```
ply
format ascii 1.0
comment PCL generated
element vertex 3
property float x
property float y
property float z
property uchar red
property uchar green
property uchar blue
element camera 1
property float view_px
property float view_py
property float view_pz
property float x_axisx
property float x_axisy
property float x_axisz
property float y_axisx
property float y_axisy
property float y_axisz
property float z_axisx
property float z_axisy
property float z_axisz
property float focal
property float scalex
property float scaley
property float centerx
property float centery
property int viewportx
property int viewporty
property float k1
property float k2
end_header
127.97639 -267.0896 -900.87177 66 107 172
128.3494 -267.11069 -900.89008 65 103 172
128.06009 -267.20074 -901.63428 68 120 175
0 0 0 1 0 0 0 1 0 0 0 1 0 0 0 0 0 3 1 0 0
```

If we add a comment at the end as shown in test3.ply it also fails
```
ply
format ascii 1.0
comment PCL generated
element vertex 3
property float x
property float y
property float z
property uchar red
property uchar green
property uchar blue
element camera 1
property float view_px
property float view_py
property float view_pz
property float x_axisx
property float x_axisy
property float x_axisz
property float y_axisx
property float y_axisy
property float y_axisz
property float z_axisx
property float z_axisy
property float z_axisz
property float focal
property float scalex
property float scaley
property float centerx
property float centery
property int viewportx
property int viewporty
property float k1
property float k2
end_header
127.97639 -267.0896 -900.87177 66 107 172
128.3494 -267.11069 -900.89008 65 103 172
128.06009 -267.20074 -901.63428 68 120 175
0 0 0 1 0 0 0 1 0 0 0 1 0 0 0 0 0 3 1 0 0
comment test
```

Even if we add another empty line after the comment it still fails as shown in test4.ply
```
ply
format ascii 1.0
comment PCL generated
element vertex 3
property float x
property float y
property float z
property uchar red
property uchar green
property uchar blue
element camera 1
property float view_px
property float view_py
property float view_pz
property float x_axisx
property float x_axisy
property float x_axisz
property float y_axisx
property float y_axisy
property float y_axisz
property float z_axisx
property float z_axisy
property float z_axisz
property float focal
property float scalex
property float scaley
property float centerx
property float centery
property int viewportx
property int viewporty
property float k1
property float k2
end_header
127.97639 -267.0896 -900.87177 66 107 172
128.3494 -267.11069 -900.89008 65 103 172
128.06009 -267.20074 -901.63428 68 120 175
0 0 0 1 0 0 0 1 0 0 0 1 0 0 0 0 0 3 1 0 0
comment test

```

I can open all 4 scenarios without a problem in meshlab, and I think it should be reasonable to expect PCL to be able to handle this. For the 3 failure cases, the error message is all the same as below

```
[pcl::PLYReader] /home/jasjuang/test2.ply:37: parse error
[pcl::PLYReader::read] problem parsing header!
[pcl::PLYReader::read] problem parsing header!
```

When I look into the source code [here](https://github.com/PointCloudLibrary/pcl/blob/a2bedc96e0b7d1a19a6e92ca9292afe009251ae5/io/src/ply/ply_parser.cpp), there are multiple places that will trigger "parse error", can someone give me pointers on where should I look into in order to fix this problem?

### PR DESCRIPTION
**Pull request by kunaltyagi**
Better error messages, better loops, less indentation.

Supersedes and closes #3527 . Fixes #3487. Closes #3353

### CODE REVIEW

**Comment by kunaltyagi** [comment]
There's a `fail` check at the end, and I don't know if I should remove that. The sole reason why that check would fail is if it has less elements, or each element has less properties than indicated. But those both are handled.

Ofc, comments inside the data are not handled, but this patch passes the test created by @jasjuang

**Comment by kunaltyagi** [comment]
Test 110: test_non_linear is flaky on my system

**Comment by kunaltyagi** [comment]
> so i think pcl can improve the fault tolerance. corlor or position loss may replace by default

https://github.com/PointCloudLibrary/pcl/pull/3353#issuecomment-534593317 https://github.com/PointCloudLibrary/pcl/pull/3527#issuecomment-572593016

With better error messages, it'll be easier to understand what's wrong. Making PCL more fault tolerant will require:
* Zero as default: depends on UI. 0, ie: black (for color) or origin (for points) as missing doesn't make universal sense. Needs change in the API to accommodate "missing value"
* Skip (and warn) items with missing properties: if it reads next element properties for current element that's cause for trouble. Needs setting in API to make it a conscious choice by the user 

PCL is not a UI tool and is also used in cases where there's near zero user feedback. As such, it needs to ensure that the input is perfect (or bail out with an error) and none of the above options can achieve that (without more extensive changes). Missing fields are a sign of a larger error and dealing with them will only cause more issues for the rest of the input file. A best effort reader will require configs and options and that'll take more effort.

**TLDR**: Not against the proposal for a best-effort reader, just against that in the current provided API.

File: io/src/ply/ply_parser.cpp [hunk 1]
```diff
@@ -99,465 +97,396 @@ bool pcl::io::ply::ply_parser::parse (const std::string& filename)
...
       }
     }
     istream >> std::ws;
-    if (istream.fail () || !istream.eof () || istream.bad ())
+    if (istream.fail ())
     {
-      if (error_callback_)
-        error_callback_ (line_number_, "parse error");
-      return false;
+      warning_callback_ (line_number_, "parse error: No newline at the end of file");
```
Comment (taketwo) [inline io/src/ply/ply_parser.cpp]: 0. Are we sure that failbit necessarily means "no newline at end"? Could there be other situations that lead to failbit being set?
1. Probably not "parse error" if we are triggering warning callback
2. `No` -> `no`
3. Do we want to report this at all?

Comment (kunaltyagi) [reply io/src/ply/ply_parser.cpp]: * 0. [This](https://en.cppreference.com/w/cpp/io/ios_base/iostate) is comprehensive. It says failbit is set when the operation to get data from input stream fails to extract correctly, ie.: can't read a char, time, number, etc. based on the overload used. Since we tried extracting white-space before, the only way it would fail is if there were no whitespace chars.
* 1 & 2. Made a mistake there, fix incoming
* 3 Yes, since line ending are required for all lines in non-empty files. See [definition of text file](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap03.html#tag_03_392)

File: io/include/pcl/io/ply/ply_parser.h [hunk 2]
```diff
@@ -383,18 +383,23 @@ namespace pcl
...
+          magic_callback_type magic_callback_ = [](){};
+          format_callback_type format_callback_ = [](format_type, const std::string&){};
+          comment_callback_type comment_callback_ = [](const std::string&){};
+          obj_info_callback_type obj_info_callback_ = [](const std::string&){};
+          end_header_callback_type end_header_callback_ = [](){return true;};
+
+          element_definition_callback_type element_definition_callbacks_ = 
+              [](const std::string&, std::size_t)
+              {
+                  return std::make_tuple([](){}, [](){});
```
Comment (taketwo) [inline io/include/pcl/io/ply/ply_parser.h]: :boom: :laughing:

Comment (kunaltyagi) [reply io/include/pcl/io/ply/ply_parser.h]: I tried fitting in more brackets but it wouldn't compile 🤣
